File: app/src/main/java/com/xnote/app/domain/markdown/RichNoteMarkdown.kt
package com.xnote.app.domain.markdown

import com.xnote.app.domain.document.InlineRun
import com.xnote.app.domain.document.ListMarker
import com.xnote.app.domain.document.NoteDocument
import com.xnote.app.domain.document.ParagraphStyle
import com.xnote.app.domain.document.TableBlock
import com.xnote.app.domain.document.TextBlock
import com.xnote.app.domain.document.plainText

// -- Constants

private val MarkdownEscapableCharacters = setOf(
    '\\', '`', '*', '_', '{', '}', '[', ']', '<', '>', '(', ')', '#', '+', '-', '.', '!', '|', '~', '=',
)

// -- Functions

fun richNoteMarkdown(title: String, document: NoteDocument): String {
    val parts = buildList {
        title.trim().takeIf { it.isNotEmpty() }?.let { value ->
            add("# ${escapeMarkdownText(value)}")
        }
        document.blocks.forEach { block ->
            when (block) {
                is TextBlock -> add(block.toMarkdown())
                is TableBlock -> add(block.toMarkdown())
                else -> error("Media blocks must be removed before Markdown conversion")
            }
        }
    }
    return parts.joinToString(separator = "\n\n").trimEnd()
}

private fun TextBlock.toMarkdown(): String {
    if (paragraphStyle == ParagraphStyle.Monospace) {
        return monospaceMarkdown(inlines.plainText(), indent, quoted)
    }

    val content = buildString {
        append(
            when (listMarker) {
                ListMarker.None -> ""
                ListMarker.Bullet, ListMarker.Dash -> "- "
                ListMarker.Numbered -> "1. "
                ListMarker.Checklist -> if (checked) "- [x] " else "- [ ] "
            },
        )
        append(
            when (paragraphStyle) {
                ParagraphStyle.Heading -> "## "
                ParagraphStyle.Subheading -> "### "
                ParagraphStyle.Body, ParagraphStyle.Monospace -> ""
            },
        )
        append(inlines.joinToString(separator = "") { it.toMarkdown() })
    }
    val prefix = "  ".repeat(indent.coerceAtLeast(0)) + if (quoted) "> " else ""
    return content.lines().joinToString(separator = "\n") { line -> prefix + line }
}

private fun TableBlock.toMarkdown(): String {
    val columnCount = rows.maxOfOrNull { it.cells.size }?.coerceAtLeast(1) ?: 1
    val normalizedRows = if (rows.isEmpty()) listOf(emptyList()) else rows.map { it.cells }
    val lines = buildList {
        add(tableRowMarkdown(normalizedRows.first(), columnCount))
        add("| ${List(columnCount) { "---" }.joinToString(" | ")} |")
        normalizedRows.drop(1).forEach { row -> add(tableRowMarkdown(row, columnCount)) }
    }
    return lines.joinToString(separator = "\n")
}

private fun tableRowMarkdown(
    cells: List<com.xnote.app.domain.document.TableCell>,
    columnCount: Int,
): String {
    val values = List(columnCount) { index ->
        cells.getOrNull(index)?.inlines
            ?.joinToString(separator = "") { it.toMarkdown() }
            ?.replace("\n", "<br>")
            .orEmpty()
    }
    return "| ${values.joinToString(" | ")} |"
}

private fun InlineRun.toMarkdown(): String {
    var value = escapeMarkdownText(text)
    if (value.isEmpty()) return value
    if (bold) value = "**$value**"
    if (italic) value = "*$value*"
    if (strikethrough) value = "~~$value~~"
    if (underline) value = "<u>$value</u>"
    if (highlight) value = "==$value=="
    if (!linkUrl.isNullOrBlank()) value = "[$value](${escapeMarkdownUrl(linkUrl)})"
    return value
}

private fun monospaceMarkdown(text: String, indent: Int, quoted: Boolean): String {
    val prefix = "  ".repeat(indent.coerceAtLeast(0)) + if (quoted) "> " else ""
    if ('\n' !in text) {
        val longestDelimiter = Regex("`+").findAll(text).maxOfOrNull { it.value.length } ?: 0
        val delimiter = "`".repeat((longestDelimiter + 1).coerceAtLeast(1))
        val padding = if (text.startsWith('`') || text.endsWith('`')) " " else ""
        return "$prefix$delimiter$padding$text$padding$delimiter"
    }
    return buildString {
        append(prefix)
        append("```\n")
        text.lines().forEachIndexed { index, line ->
            if (index > 0) append('\n')
            append(prefix)
            append(line)
        }
        append('\n')
        append(prefix)
        append("```")
    }
}

private fun escapeMarkdownText(text: String): String = buildString(text.length) {
    text.forEach { character ->
        if (character in MarkdownEscapableCharacters) append('\\')
        append(character)
    }
}

private fun escapeMarkdownUrl(url: String): String = buildString(url.length) {
    url.trim().forEach { character ->
        if (character == '\\' || character == ')') append('\\')
        append(character)
    }
}

File: flutter/lib/domain/markdown/rich_note_markdown.dart
import '../document/note_block.dart';
import '../document/note_document.dart';

// -- Constants

const _markdownEscapableCharacters = <String>{
  r'\',
  '`',
  '*',
  '_',
  '{',
  '}',
  '[',
  ']',
  '<',
  '>',
  '(',
  ')',
  '#',
  '+',
  '-',
  '.',
  '!',
  '|',
  '~',
  '=',
};

// -- Functions

String richNoteToMarkdown(String title, NoteDocument document) {
  final parts = <String>[];
  final trimmedTitle = title.trim();
  if (trimmedTitle.isNotEmpty) {
    parts.add('# ${_escapeMarkdownText(trimmedTitle)}');
  }
  for (final block in document.blocks) {
    switch (block) {
      case TextBlock():
        parts.add(_textBlockToMarkdown(block));
      case TableBlock():
        parts.add(_tableBlockToMarkdown(block));
      case ImageBlock() || StickerBlock() || DrawingBlock():
        throw StateError(
            'Media blocks must be removed before Markdown conversion');
    }
  }
  return parts.join('\n\n').trimRight();
}

String _textBlockToMarkdown(TextBlock block) {
  if (block.paragraphStyle == ParagraphStyle.monospace) {
    return _monospaceMarkdown(
      plainText(block.inlines),
      block.indent,
      block.quoted,
    );
  }

  final marker = switch (block.listMarker) {
    ListMarker.none => '',
    ListMarker.bullet || ListMarker.dash => '- ',
    ListMarker.numbered => '1. ',
    ListMarker.checklist => block.checked ? '- [x] ' : '- [ ] ',
  };
  final heading = switch (block.paragraphStyle) {
    ParagraphStyle.heading => '## ',
    ParagraphStyle.subheading => '### ',
    ParagraphStyle.body || ParagraphStyle.monospace => '',
  };
  final content =
      '$marker$heading${block.inlines.map(_inlineToMarkdown).join()}';
  final prefix = '${_repeat('  ', block.indent)}${block.quoted ? '> ' : ''}';
  return content.split('\n').map((line) => '$prefix$line').join('\n');
}

String _tableBlockToMarkdown(TableBlock block) {
  final columnCount = block.columnCount < 1 ? 1 : block.columnCount;
  final rows = block.rows.isEmpty ? <TableRow>[TableRow()] : block.rows;
  final lines = <String>[
    _tableRowToMarkdown(rows.first, columnCount),
    '| ${List<String>.filled(columnCount, '---').join(' | ')} |',
    for (final row in rows.skip(1)) _tableRowToMarkdown(row, columnCount),
  ];
  return lines.join('\n');
}

String _tableRowToMarkdown(TableRow row, int columnCount) {
  final values = <String>[
    for (var index = 0; index < columnCount; index += 1)
      if (index < row.cells.length)
        row.cells[index].inlines
            .map(_inlineToMarkdown)
            .join()
            .replaceAll('\n', '<br>')
      else
        '',
  ];
  return '| ${values.join(' | ')} |';
}

String _inlineToMarkdown(InlineRun inline) {
  var value = _escapeMarkdownText(inline.text);
  if (value.isEmpty) {
    return value;
  }
  if (inline.bold) {
    value = '**$value**';
  }
  if (inline.italic) {
    value = '*$value*';
  }
  if (inline.strikethrough) {
    value = '~~$value~~';
  }
  if (inline.underline) {
    value = '<u>$value</u>';
  }
  if (inline.highlight) {
    value = '==$value==';
  }
  if (inline.linkUrl?.trim().isNotEmpty ?? false) {
    value = '[$value](${_escapeMarkdownUrl(inline.linkUrl!)})';
  }
  return value;
}

String _monospaceMarkdown(String text, int indent, bool quoted) {
  final prefix = '${_repeat('  ', indent)}${quoted ? '> ' : ''}';
  if (!text.contains('\n')) {
    final delimiters = RegExp(r'`+').allMatches(text);
    var longestDelimiter = 0;
    for (final match in delimiters) {
      if (match.group(0)!.length > longestDelimiter) {
        longestDelimiter = match.group(0)!.length;
      }
    }
    final delimiter = _repeat('`', longestDelimiter + 1);
    final padding = text.startsWith('`') || text.endsWith('`') ? ' ' : '';
    return '$prefix$delimiter$padding$text$padding$delimiter';
  }
  final lines = text.split('\n').map((line) => '$prefix$line').join('\n');
  return '$prefix```\n$lines\n$prefix```';
}

String _escapeMarkdownText(String text) {
  final output = StringBuffer();
  for (final codePoint in text.runes) {
    final character = String.fromCharCode(codePoint);
    if (_markdownEscapableCharacters.contains(character)) {
      output.write(r'\');
    }
    output.write(character);
  }
  return output.toString();
}

String _escapeMarkdownUrl(String url) {
  final output = StringBuffer();
  for (final codePoint in url.trim().runes) {
    final character = String.fromCharCode(codePoint);
    if (character == r'\' || character == ')') {
      output.write(r'\');
    }
    output.write(character);
  }
  return output.toString();
}

String _repeat(String value, int count) =>
    List<String>.filled(count, value).join();
